asm/instruction.go

```diff
@@ -299,16 +318,68 @@ func (ins Instruction) Format(f fmt.State, c rune) {
 	}
 
 ref:
-	if ins.Reference != "" {
-		fmt.Fprintf(f, " <%s>", ins.Reference)
+	if ins.Reference() != "" {
+		fmt.Fprintf(f, " <%s>", ins.Reference())
 	}
 }
 
+func (ins Instruction) equal(other Instruction) bool {
```

Comment:
This should also compare `metadata` to match the old behaviour.

Reply:
We have gone back and forth on this already with @lmb

https://github.com/cilium/ebpf/pull/567#discussion_r805953880
> I'm not sure whether this should be part of equality. Does an instruction change because it has a different line number? Seems too specific to our tests to be of general use. I'd prefer not exporting `Equal` at all.

We first need some conciseness, before proceeding. 

From my point of view, the equals function is only used by our unit tests. When would a user of the library need to check for equality? So to me, having an un-exported equal makes sense. But perhaps @ti-mo is more familiar with how ebpf is used in other projects and if equality checking is something users need.

Reply:
https://github.com/cilium/ebpf/pull/583 means `instructionComparer` is no longer necessary, let's keep `equal()` unexported.